File: 0437-path-sum-iii/0437-path-sum-iii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    long answer = 0;
    int mod = (1e9 + 7);
    void kSumPaths(TreeNode* root, int k, long long prefixSum, unordered_map<long, long> &mp){
        if(!root){
            return;
        }
        
        prefixSum += root->val;
        if(mp.find(prefixSum - k) != mp.end()){
            answer = (answer + mp[prefixSum - k]) % mod;
        }
        
        mp[prefixSum]++;
        kSumPaths(root->left, k, prefixSum, mp);
        kSumPaths(root->right, k, prefixSum, mp);
        mp[prefixSum]--;
    }

    int pathSum(TreeNode* root, int targetSum) {
        unordered_map<long, long> mp;
        mp[0] = 1;
        kSumPaths(root, targetSum, 0, mp);
        return answer;
    }
};
        

File: 0437-path-sum-iii/0437-path-sum-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        result = [0]
        prefixMap = {0: 1}
        def pathSumK(root, currentSum):
            if root == None:
                return

            currentSum += root.val

            if prefixMap.get(currentSum - targetSum) != None:
                result[0] += prefixMap.get(currentSum - targetSum)
            

            prefixMap[currentSum] = prefixMap.get(currentSum, 0) + 1
            pathSumK(root.left, currentSum)
            pathSumK(root.right, currentSum)
            prefixMap[currentSum] -= 1
        
        pathSumK(root, 0)
        return result[0]
        